pkg/core/tekton/variables_resolver.go

```diff
@@ -0,0 +1,51 @@
+package tekton
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+type variablesResolver struct {
+	references map[string]string
+}
+
+func newVariablesResolver() *variablesResolver {
+	r := &variablesResolver{}
+	r.references = make(map[string]string)
+	return r
+}
+
+func (r *variablesResolver) resolve(value string) string {
+	regex := regexp.MustCompile(`{([^{{{}}}]*)}`)
+	matches := regex.FindAllStringSubmatch(value, -1)
+	if len(matches) == 0 {
+		if v, ok := r.references[value]; ok {
+			return v
+		}
+	}
+	for _, v := range matches {
+		toResolve := strings.TrimSpace(v[1])
+		resolvedTo, existsReference := r.references[toResolve]
+		// if value begins with 'steps.' it could mean that the value is derived from a
+		// task output, in that case transalate it to the tekton way of getting a task
+		// output "$(tasks.TASK.results.VARIABLE), or that the value is referencing a step
+		// input/output that is resolvable here."
```

Comment:
Could you add some specific examples here? For example that trainer image evaluation...

Reply:
done.